docs/enhancements/install-strategy.md

```diff
@@ -0,0 +1,196 @@
+# Hive Install Strategy API Refactor
+
+## Summary
+
+Break install strategy out of ClusterDeployment.Spec and allow for multiple CRD implementations that adhere to an interface, where Hive interacts with these CRDs dynamically via duck typing.
+
+
+## Motivation
+
+Hive is presently struggling with how to design the API to meet the needs of provisioning standalone, agent based, and hypershift clusters. The current code base evolved quite specific to running openshift-install pods and has some growing pains to accomodate new paths to provision clusters. Changes for one raise a lot of questions and inconsistencies for the others as everything is lumped into ClusterDeployment. The line between day 1 and day 2 is very unclear and mostly just enforced via increasingly complicated validating webhooks.
+
+### Goals
+
+- Clarify Hive APIs. (day 1 vs day 2)
+- Provide space for implementations to define the spec and status they need without causing uncomfortable complications for other implementations.
+- Breakout AgentInstallStrategy to a separate CRD. This is still guarded by an alpha feature gate and thus we should still be able to move it around.
+- Do not break API compatability for existing standalone OpenShift cluster provisioning. We hope to slowly migrate those installs to this system by offering it in parallel and eventually deprecating the old, but for now the focus is on breaking out the AgentInstallStrategy.
+
+### Non-Goals
```

Comment:
how do we implement cluster pools for these new ClusterInstall implementations? Would that be a separate discussion? Or will some of the decisions we make here be influenced by cluster pool implementation...

Reply:
Also can we include details on the deprovision of the clusters. I would be in favor of implementing the deprovision as the delete of the object that implements the ClusterInstall instead of a new different object.

we shuoldn't need to support a deprovision that was not provisioned by our APIs.

Reply:
I have not thought about ClusterPool yet, will give that a look and update!

Historically at one point, OSD was supporting a provision a cluster feature where the CLusterDeployment was then removed, and the user was totally in control of the cluster after that point. Then, they wanted to be able to optionally deprovision those, which led us to an original implementation of ClusterDeprovision that was not tied to any ClusterDeployment as it may not exist. This entire feature eventually died, I'm not aware of anyone needing us to deprovision a cluster we did not provision anymore.

It feels like your suggestion would be safe, move the deprovision finalizer to the ClusterInstall.

Reply:
ClusterPool support is an interesting one. Looking at agent install strategy as it sits today, it *looks* like you might be able to get away with saying that an install strategy should be re-usable by multiple clusters? (can networking be shared?) However I really doubt this will be generically true. If it's not true, does it even make sense to support ClusterPool's for that platform/clusterinstall impl? I'm not sure it does. 

If we require specific complex settings per-cluster, ClusterPool is a very tough target to hit, it's basically why we're struggling with that vmware + arbitrary DNS requirement.

Reply:
Actually now I'm thinking this might be the key to solving that vmware DNS thing. ClusterPool.Spec.ClusterInstallRef points to a VMWareMyCustomDNS ClusterInstall. We don't care what or how that ClusterInstall defines or configures DNS for each cluster, it's up to whomever implements the controllers for it. Hive will just copy the VMWareMyCustomDNS ClusterInstall, as-is, directly to the ClusterDeployment namespace when we create it and associate it with the ClusterDeployment. 

Assumption is that if you're going to use it with a ClusterPool, it can't contain anything cluster specific, but it could contain ways to obtain something cluster specific. Adding a section with this proposal for your thoughts. This enhancement will list implementation of this as a non-goal.

Reply:
> Also can we include details on the deprovision of the clusters. I would be in favor of implementing the deprovision as the delete of the object that implements the ClusterInstall instead of a new different object.
> we shuoldn't need to support a deprovision that was not provisioned by our APIs.


A little unsure here now that I think about it. Wouldn't it feel like I provisioned my cluster, I have my ClusterDeployment now for day 2. I don't need my ClusterInstall anymore, so why not free up some space in etcd and delete it, which could then unexpectedly trash their cluster.

Reply:
It seems like the discussion is getting into feature overlap with central machine management, especially if openshift rebases to the latest cluster-api. It might make sense to further explore those implications before trying to solve deprovision here. If you already have an idea of what that will look like relative to hive, I'd be very interested to know more.